### ISSUE TITLE
Add networkpolicy

### ISSUE THREAD

**Issue by devopstales** [oauth2-proxy/manifests#103]
I tried to use the oauth proxy in a cluster where networkpolicy is enabled. The pod can not start and not logging anything.

I have the fallowing policies: 

```
---
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  name: allow-internet-only
  namespace: ingress-system
spec:
  podSelector:
    matchLabels:
      app.kubernetes.io/component: authentication-proxy
      app.kubernetes.io/instance: oauth2-proxy
  policyTypes:
  - Egress
  egress:
  - to:
    - ipBlock:
        cidr: 0.0.0.0/0
---
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  name: allow-to-oauth2-proxy--ingress-nginx-private--controller
  namespace: ingress-system
spec:
  podSelector:
    matchLabels:
      app.kubernetes.io/component: controller
      app.kubernetes.io/instance: ingress-nginx-private
  policyTypes:
  - Egress
  egress:
  - to:
    - podSelector:
        matchLabels:
          app.kubernetes.io/component: authentication-proxy
          app.kubernetes.io/instance: oauth2-proxy
```

**Comment by pierluigilenoci** [comment]
@devopstales with this kind of information it is not possible to tell what is wrong with the configuration.
Are there any policies that deny connections?

**Comment by pierluigilenoci** [comment]
@devopstales any update on this?

**Comment by devopstales** [comment]
@pierluigilenoci  I use cilium for my networking and this is all of my policies. Logically for me what is not allowed that is blocked. So anything other than this is blocked. I didn't know how oauth2-proxy works in the low level. I created this policies with best of my knowledge. I think main problem is the  helm chart lacks network policies.

**Comment by pierluigilenoci** [comment]
@devopstales let me make a premise: the Network Policies are completely external to the software and therefore not dependent on the software itself or on how it is installed. 

For this reason, OAuth2-Proxy is capable of working with NetworkPolicy configured within clusters.
Or if they are not.

NPs only change the way traffic enters and exits the pod.
Obviously, if the NPs block traffic in some way the pod does not work correctly but this is not the responsibility of the pod but of the policies.

Furthermore, considering how "personal" the configuration of the policies is for each individual case, I don't think it is possible or reasonable to include them in the OAuth2-Proxy chart but it makes sense to do it externally.

The chart, however, provides for the possibility of adding extra manifest, and therefore you can add all the policies that are right for you without problems. 
Ref:  https://github.com/oauth2-proxy/manifests/blob/main/helm/oauth2-proxy/values.yaml#L297

What I can suggest is to study in depth how the Network Policies are implemented by Cillium and configure them accordingly.
Ref:
https://docs.cilium.io/en/latest/security/policy/intro/
https://kubernetes.io/docs/tasks/administer-cluster/network-policy-provider/cilium-network-policy/

I have no experience with Cillium but there are some points that I can still suggest you check:
- are you using the right API? `networking.k8s.io/v1` or `cilium.io/v2`
- if not, are you using the right manifest? `NetworkPolicy ` or `CiliumNetworkPolicy `
- in your example you have opened the connection from OAuth2-Proxy to the outside (so the pod can connect to the internet) but is it open in ingress? I believe you need to configure an NP that allows an external pod (e.g. Nginx) to connect to OAuth2-Proxy.
- you have opened the Nginx pod to connect to OAuth2-Proxy but the controller is not open for third-party connection. So eventually a load-balancer would be able to open a connection to Nginx?

In summary, check that you have used the right manifest and remember that "each NP" is made up of egress and ingress and they are both important. If Cillium blocks everything you have to make them both explicit.

**Comment by pierluigilenoci** [comment]
Small addendum for labels: the label `app.kubernetes.io/instance` could lead to problems related to the fact that it is connected to the `.Release.Name` of the chart and therefore is subject to change. 

I suggest using these:
```yaml
    app.kubernetes.io/component: authentication-proxy
    app.kubernetes.io/name: oauth2-proxy
```

**Comment by devopstales** [comment]
@pierluigilenoci Dose not mater what application we are talking about, If the application cannot connect where it need to, it wont work. You can not expect the end user to understand  the app as those who created them. That is why most application's chart contains basic NPs, to grantee the correct working, or at least grantee the app will start. What can a user do if the app didn't started? Read the logs! But there is no logs!!! If the app cannot connect to somewhere, at least log it. Of course every situation is different, and the users responsibility to modify for there use case, but it is the responsibility  of the creators to add sufficient information to do this. Now the main problem is not that the authentication not working but the pod isn't starting.

**Comment by pierluigilenoci** [comment]
@devopstales let's start by specifying that I'm not one of the OAuth2-Proxy developers but only one of the Chart maintainers.
The developers can be found here: https://github.com/oauth2-proxy/oauth2-proxy

Furthermore, this is an open-source project and any responsibility for the project lies with the entire community, certainly not only with those who develop it.
Developers are usually just people of goodwill who make themselves available to others.
Without being paid.
If you want someone to do something as you please, the only solution is to pay for it.

I would like to add that if you have to use any software, you are not required to know (at least initially) how it works "inside" but you certainly need to know how it interacts with the outside world.

So that you learn roughly at least the minimum interactions that OAuth2-Proxy has with any software that needs authentication and any Auth provider seems to be the minimum.

Having said that, the software under discussion only receives calls on the service port that is opened and configured with these options:
https://github.com/oauth2-proxy/manifests/blob/main/helm/oauth2-proxy/values.yaml#L100-L108 

And in any case, it is configured using these options (which you can possibly change via the chart):
https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview/

To return to the specific case of Network policies it is necessary to take into account that Kubernetes supports different network drivers and different plugins. 

So there are multiple ways to implement Network Policies that are completely dependent on how the cluster is configured and are completely independent of how the software you run works.

Considering that it would not make sense to cover all the possible permutations of configurations and it would not even be possible to make a more generic one because each cloud provider uses a different configuration, I do not see what advantage there can be to add something that only works for a small percentage of cases.

I can understand that you need a solution to your specific problem but from what you write the root is far from the network policies.

For example, the fact that the pod doesn't even produce a single line of logs is at least peculiar.

I can assure you that the OAuth2-Proxy pod, regardless of the network configuration, or regardless of any configuration, starts and produces logs in any case. Maybe it goes into CrashLoopBackOff but it produces something.

Can you access the node where the pod runs? 
Can you see if the logs are at least written to disk in the node? 
If you try to `describe` the OAuth2-Proxy pod, what state is the pod in?

Despite the tone you use which is not particularly accommodating, even if I want to be of goodwill with the little information you have made available it is very difficult to help you.

If you believe that the lack of Network Policies is very serious, nothing prevents you from implementing them yourself and doing a PR towards this repository. 

I promise you a quick review and integration process.

**Comment by pierluigilenoci** [comment]
@devopstales any update on this?

**Comment by pierluigilenoci** [comment]
It was closed due to lack of feedback from the reporter.
